### PR TITLE
Metempsychosis Returns

### PR DESCRIPTION
**Pull request by VMSolidus**
## About the PR
This PR brings back a much beloved feature from Nyanotrasen, the Metempsychosis Machine. Metempsychosis is a Nyanotrasen variant on the cloning machine. Instead of giving you a new body, it forcibly reincarnates its patient(victim) into an entirely new body. 

## Why / Balance
Back by INCREDIBLY POPULAR demand. 

## Technical details
So, technically speaking, Wizden no longer maintains the Cloning code. It is effectively gone from their codebase, and no longer present on their servers. This realistically means that the absurd amount of changes required to make this happen will essentially never be an issue. 

## Media
image

**Changelog**

:cl: VMSolidus
- add: Metempsychosis machines have been added to the game.

### CODE REVIEW

**Comment by VMSolidus** [comment]
image
And now the first of the two midround antags that required Metempsychosis are back. I've also fixed some bugs with Metempsychosis. It no longer randomizes character name on the 2nd reincarnation onwards.

**Review by DebugOk — CHANGES_REQUESTED** [review]
This is adding 2 unrelated features that aren't part of metempsychosis.

**Comment by VMSolidus** [comment]
> This is adding 2 unrelated features that aren't part of metempsychosis.

OK. Fine. They've been reverted. This now only includes code for Metempsychosis. I'll just wait until this has been merged before I do anything with Fugitive and Paradox Anomaly. Both require this as a prerequisite.

File: Content.Server/Cloning/CloningSystem.cs
Comment (DEATHB4DEFEAT) [inline Content.Server/Cloning/CloningSystem.cs]: Codebase marker

File: Content.Server/Medical/Components/MedicalScannerComponent.cs
Comment (DEATHB4DEFEAT) [inline Content.Server/Medical/Components/MedicalScannerComponent.cs]:
```suggestion

        // Nyano, needed for Metem Machine.
```

File: Content.Server/Nyanotrasen/Cloning/MetempsychosisKarmaComponent.cs
Comment (DEATHB4DEFEAT) [inline Content.Server/Nyanotrasen/Cloning/MetempsychosisKarmaComponent.cs]:
```suggestion
namespace Content.Server.Nyanotrasen.Cloning
```

File: Content.Server/Nyanotrasen/Cloning/MetempsychoticMachineComponent.cs
Comment (DEATHB4DEFEAT) [inline Content.Server/Nyanotrasen/Cloning/MetempsychoticMachineComponent.cs]:
```suggestion
namespace Content.Server.Nyanotrasen.Cloning
```

File: Content.Server/Nyanotrasen/Cloning/MetempsychoticMachineSystem.cs
Comment (DEATHB4DEFEAT) [inline Content.Server/Nyanotrasen/Cloning/MetempsychoticMachineSystem.cs]:
```suggestion
namespace Content.Server.Nyanotrasen.Cloning
```

File: Content.Server/Nyanotrasen/Cloning/MetempsychoticMachineSystem.cs
Comment (DEATHB4DEFEAT) [inline Content.Server/Nyanotrasen/Cloning/MetempsychoticMachineSystem.cs]: I don't know the names but I believe there's prototype validator attributes you should add to these

File: Content.Server/Nyanotrasen/Cloning/MetempsychoticMachineSystem.cs
Comment (DEATHB4DEFEAT) [inline Content.Server/Nyanotrasen/Cloning/MetempsychoticMachineSystem.cs]:
```suggestion
            if (karma != null)
                chance -= ((1 - component.HumanoidBaseChance) * (float) karma);
```

File: Content.Server/Nyanotrasen/Cloning/MetempsychoticMachineSystem.cs
Comment (DEATHB4DEFEAT) [inline Content.Server/Nyanotrasen/Cloning/MetempsychoticMachineSystem.cs]:
```suggestion
            if (chance > 1 && _random.Prob(chance - 1))
            {
                species = oldSpecies;
                return oldSpecies.Prototype;
            }
            else
                chance = 1;
```

File: Content.Server/Nyanotrasen/Cloning/MetempsychoticMachineSystem.cs
Comment (DEATHB4DEFEAT) [inline Content.Server/Nyanotrasen/Cloning/MetempsychoticMachineSystem.cs]:
```suggestion
            if (_random.Prob(chance) &&
                _prototypeManager.TryIndex<WeightedRandomPrototype>(MetempsychoticHumanoidPool, out var humanoidPool) &&
                _prototypeManager.TryIndex<SpeciesPrototype>(humanoidPool.Pick(), out var speciesPrototype))
            {
                species = speciesPrototype;
                return speciesPrototype.Prototype;
            }
            else
            {
                species = null;
                Logger.Error("Could not index species for metempsychotic machine...");
                return "MobHuman";
            }
```

File: Resources/Prototypes/Nyanotrasen/Entities/Structures/Machines/metempsychoticMachine.yml
Comment (DEATHB4DEFEAT) [inline Resources/Prototypes/Nyanotrasen/Entities/Structures/Machines/metempsychoticMachine.yml]:
```suggestion
    - type: MetempsychoticMachine
    - type: CloningPod
    - type: Machine
      board: MetempsychoticMachineCircuitboard
    - type: Sprite
      sprite: Nyanotrasen/Structures/Machines/metempsychotic.rsi
      snapCardinals: true
      layers:
        - state: pod_0
    - type: Appearance
    - type: GenericVisualizer
      visuals:
        enum.CloningPodVisuals.Status:
          base:
            Cloning: { state: pod_1 }
            NoMind: { state: pod_1 }
            Gore: { state: pod_1 }
            Idle: { state: pod_0 }
    - type: Psionic
```

File: Resources/Prototypes/Nyanotrasen/Research/biochemical.yml.yml
Comment (DEATHB4DEFEAT) [inline Resources/Prototypes/Nyanotrasen/Research/biochemical.yml.yml]:
```suggestion
    - BiomassReclaimerMachineCircuitboard
    - CloningConsoleComputerCircuitboard
    - MedicalScannerMachineCircuitboard
    - MetempsychoticMachineCircuitboard
```

File: Resources/Textures/Nyanotrasen/Structures/Machines/metempsychotic.rsi/meta.json
Comment (DEATHB4DEFEAT) [inline Resources/Textures/Nyanotrasen/Structures/Machines/metempsychotic.rsi/meta.json]:
```suggestion
      "delays": [ [ 0.1, 0.1, 0.1, 0.1, 0.1, 0.1 ] ]
```

**Review by DEATHB4DEFEAT — CHANGES_REQUESTED** [review]

File: Content.Server/Cloning/CloningSystem.cs
Comment (VMSolidus) [reply Content.Server/Cloning/CloningSystem.cs]: Got it

File: Content.Server/Nyanotrasen/Cloning/MetempsychoticMachineSystem.cs
Comment (VMSolidus) [reply Content.Server/Nyanotrasen/Cloning/MetempsychoticMachineSystem.cs]: I genuinely don't know what a prototype validator is or what you're referring to? Or why that would be an issue.

**Comment by VMSolidus** [comment]
image
Test fails are illegitimate, ubuntu-latest is just trying to gaslight me.

File: Content.Server/Nyanotrasen/Cloning/MetempsychoticMachineSystem.cs
Comment (deltanedas) [reply Content.Server/Nyanotrasen/Cloning/MetempsychoticMachineSystem.cs]: `[ValidatePrototypeId<WhateverPrototype>]`

basically the issue is if you rename or remove the prototype the code would still assume it exists and crash somewhere, with validator you can catch that early at build time

Comment (VMSolidus) [reply Content.Server/Nyanotrasen/Cloning/MetempsychoticMachineSystem.cs]: Got it

File: Content.Server/Cloning/CloningConsoleSystem.cs
Comment (DEATHB4DEFEAT) [inline Content.Server/Cloning/CloningConsoleSystem.cs:35]:
```suggestion

```

File: Content.Server/Cloning/CloningConsoleSystem.cs
Comment (DEATHB4DEFEAT) [inline Content.Server/Cloning/CloningConsoleSystem.cs]:
```suggestion
            // Nyano: Adds scannerComp.MetemKarmaBonus
            if (_cloningSystem.TryCloning(cloningPodUid, body.Value, (mindId, mind), cloningPod, scannerComp.CloningFailChanceMultiplier, scannerComp.MetemKarmaBonus))
```

File: Content.Server/Cloning/CloningSystem.cs
Comment (DEATHB4DEFEAT) [inline Content.Server/Cloning/CloningSystem.cs]: Single line block comment and an uncommented change?

File: Content.Server/Cloning/CloningSystem.cs
Comment (DEATHB4DEFEAT) [inline Content.Server/Cloning/CloningSystem.cs]:
```suggestion
                    if (speciesPrototype.ID != newSpecies.ID)
                        switchingSpecies = true;
```

**Review by DEATHB4DEFEAT — CHANGES_REQUESTED** [review]

File: Content.Server/Cloning/CloningSystem.cs
Comment (DEATHB4DEFEAT) [reply Content.Server/Cloning/CloningSystem.cs]:
```suggestion
            // Begin Nyano-code: allow paradox anomalies to be cloned.
            var pref = humanoid.LastProfileLoaded;

            if (pref == null)
                return false;
            // End Nyano-code
```

File: Content.Server/Medical/Components/MedicalScannerComponent.cs
Comment (DebugOk) [inline Content.Server/Medical/Components/MedicalScannerComponent.cs]: Machine upgrading has been removed

File: Content.Shared/Humanoid/HumanoidAppearanceComponent.cs
Comment (DebugOk) [inline Content.Shared/Humanoid/HumanoidAppearanceComponent.cs]: Is there seriously no other way to do this? Surely code for it already exists?

File: Resources/Prototypes/Nyanotrasen/metempsychoticHumanoids.yml
Comment (DebugOk) [inline Resources/Prototypes/Nyanotrasen/metempsychoticHumanoids.yml:2]: Are these automatically checked to see if all prototypes are valid? If not, this'll require a test to be added

File: Resources/Prototypes/Nyanotrasen/Entities/Structures/Machines/metempsychoticMachine.yml
Comment (DebugOk) [inline Resources/Prototypes/Nyanotrasen/Entities/Structures/Machines/metempsychoticMachine.yml:25]: Why is the pod psionic?

File: Content.Server/Nyanotrasen/Cloning/MetempsychoticMachineSystem.cs
Comment (DebugOk) [inline Content.Server/Nyanotrasen/Cloning/MetempsychoticMachineSystem.cs]: Dont use logger, thats what sawmills are for

File: Content.Server/Cloning/CloningSystem.cs
Comment (DebugOk) [inline Content.Server/Cloning/CloningSystem.cs]: Why is this line changing?

**Review by DebugOk — CHANGES_REQUESTED** [review]

File: Resources/Prototypes/Nyanotrasen/Entities/Structures/Machines/metempsychoticMachine.yml
Comment (VMSolidus) [reply Resources/Prototypes/Nyanotrasen/Entities/Structures/Machines/metempsychoticMachine.yml:25]: Everything that interacts directly with the Psionic system needs to have this component. In this context it's primarily for Metapsionic Pulse, and with the refactor that I'm actively coding also allows it to have unique flavor text when scanned with a focused Metapsionic pulse. To clarify, Psionic component does not imply that an entity is itself a psychic, merely stating that "This entity interacts with the Psionic systems in some way". With the exception of messing with Mantes, it is otherwise vestigial right now, I have left this component present because the original version of it from legacy code had the component.

File: Content.Shared/Humanoid/HumanoidAppearanceComponent.cs
Comment (VMSolidus) [reply Content.Shared/Humanoid/HumanoidAppearanceComponent.cs]: This right here is the part that makes Metempsychosis a prerequisite for Fugitives and Paradox Anomalies. However since the original Metempsychosis code was written it's possible to rewrite Fugitive without this entirely. This is because HumanoidCharacterProfile systems are vastly more robust than they were, and I have access to a suite of tools for generating a character randomly or procedurally. Paradox Anomalies by contrast use it so that they can spawn as the original at-shift-start version of a character rather than any modified versions from Metem machines. There isn't any functionality given by updates to HCP systems that allow Paradox Anoms to exist without Metempsychosis. Unless of course we want Paradox Anoms to always spawn as the exact current appearance of a given character.

File: Resources/Prototypes/Nyanotrasen/metempsychoticHumanoids.yml
Comment (VMSolidus) [reply Resources/Prototypes/Nyanotrasen/metempsychoticHumanoids.yml:2]: image
Yes these are automatically validated by MetempsychoticMachineSystem.

Comment (DebugOk) [reply Resources/Prototypes/Nyanotrasen/metempsychoticHumanoids.yml:2]: By creating errors if they aren't valid yes. Nothings stopping you from comitting invalid prototypes

**Review by DebugOk — DISMISSED** [review]
Its probably fine, I didn't spot any error just reading over everything

**Review by DEATHB4DEFEAT — APPROVED** [review]

**Comment by VMSolidus** [comment]
I haven't been able to merge this yet since when Debug approved it, I required Death's approval too. Then merging Paradox anoms without Metem first caused a merge conflict, removing Debug's approval. Now I have Death's approval but not Debug's, so this still can't be merged.

**Review by DebugOk — APPROVED** [review]

**Pull request merged**
